### PR TITLE
Added UI for search History

### PR DESCRIPTION
**Pull request by saurabh95**
The UI is inspired from https://trello.com/c/IzACddEv/344-s-find-recent-search-history with minor tweaks.

Dark Theme

image

Light Theme

image

The UI can be invoked using arrow up and down keys or clicking on the down arrow on the right of the search input.
The UI is reused from QuickSearchField.js which is also used by Quick Open.

### CODE REVIEW

**Comment by ficristo** [comment]
I leave the review for someone else here.
I only have one question: when you select an item from the dropdown list and then reopen the list, only the item selected is visible. Is this expected?
If I need to change the item, I have to clear the content of the search box and only then I'm able to see other items in the dropdown list, which is a bit of a pain.

Do you still have an UI / UX designer that can review these kind changes ?

**Comment by saurabh95** [comment]
@ficristo Actually I have added a filtering in the dropdown, if there is some text in the search field and then we try to open dropdown, then only elements in dropdown which have that string as its substring would be visible.

**Comment by saurabh95** [comment]
@ficristo discussed it with Rajnish, he also feels that removing the filtering would be better, so now whole search history is visible irrespective of content in the search field

File: src/search/FindBar.js [hunk 1]
```diff
@@ -281,6 +285,17 @@ define(function (require, exports, module) {
                     self.trigger("doFind");
                 }
             })
+            .on("click", ".dropdown-icon", function (e) {
+                if (!self.searchField) {
+                    self.showSearchHints();
+                } else if ($(".quick-search-container").is(':visible')) {
+                    $(".quick-search-container").hide();
+                } else if (!$(".quick-search-container").is(':visible')) {
```
Comment (petetnt) [inline src/search/FindBar.js]: This can be just `else`, right?

File: src/search/FindBar.js [hunk 2]
```diff
@@ -324,7 +339,9 @@ define(function (require, exports, module) {
                             searchHistory.pop();
                         }
                     }
-                    searchHistory.unshift($('#find-what').val());
+                    if ($('#find-what').val()) {
+                        searchHistory.unshift($('#find-what').val());
```
Comment (petetnt) [inline src/search/FindBar.js]: Maybe cache the `#find-what` value to a variable

File: src/search/FindBar.js [hunk 3]
```diff
@@ -281,6 +285,17 @@ define(function (require, exports, module) {
                     self.trigger("doFind");
                 }
             })
+            .on("click", ".dropdown-icon", function (e) {
+                if (!self.searchField) {
+                    self.showSearchHints();
+                } else if ($(".quick-search-container").is(':visible')) {
```
Comment (petetnt) [inline src/search/FindBar.js]: `${".quick-search-container"}` could be cached to a variable

File: src/search/FindBar.js [hunk 4]
```diff
@@ -348,13 +365,11 @@ define(function (require, exports, module) {
                     }
                     historyIndex = 0;
                 } else if (e.keyCode === KeyEvent.DOM_VK_DOWN || e.keyCode === KeyEvent.DOM_VK_UP) {
-                    if (e.keyCode === KeyEvent.DOM_VK_DOWN) {
-                        historyIndex = (historyIndex - 1 + searchHistory.length) % searchHistory.length;
-                    } else {
-                        historyIndex = (historyIndex + 1 + searchHistory.length) % searchHistory.length;
+                    if (!self.searchField) {
+                        self.showSearchHints();
+                    } else if (!$(".quick-search-container").is(':visible')) {
```
Comment (petetnt) [inline src/search/FindBar.js]: Use the cached variable here too (see the previous review comment)

File: src/search/FindBar.js [hunk 5]
```diff
@@ -407,6 +422,41 @@ define(function (require, exports, module) {
     };
 
     /**
+     * @private
+     * Shows the search History in dropdown.
+     */
+    FindBar.prototype.showSearchHints = function () {
+        var self = this;
+        this.$searchField = $("input#find-what");
```
Comment (petetnt) [inline src/search/FindBar.js]: This part (L430-L432) with nested `this.$searchField` looks a bit scary, maybe refactor it to use some named variables instead.

File: src/search/FindBar.js [hunk 6]
```diff
@@ -281,6 +285,17 @@ define(function (require, exports, module) {
...
+            .on("click", ".dropdown-icon", function (e) {
+                if (!self.searchField) {
+                    self.showSearchHints();
+                } else if ($(".quick-search-container").is(':visible')) {
+                    $(".quick-search-container").hide();
+                } else if (!$(".quick-search-container").is(':visible')) {
+                    self.searchField.setText($("#find-what").val());
+                    $(".quick-search-container").show();
+                }
+                self.$("#find-what").focus();
```
Comment (petetnt) [inline src/search/FindBar.js]: What's the difference between `self.$("#find-what")` and `$("#find-what")`?

**Review by petetnt — CHANGES_REQUESTED** [review]
Added a few comments, mostly looks really great 👍

File: src/search/FindBar.js (hunk 2 above)
Comment (swmitra) [inline src/search/FindBar.js]: I second @petetnt. Value of `#find-what` should be assigned to a variable while initiating the search session. Whole session should refer to the same value for navigation.

File: src/search/FindBar.js (hunk 6 above)
Comment (saurabh95) [inline src/search/FindBar.js]: Actually there are chances of having multiple find bars present (one is closing, other might be opening), so there can be multiple `$("#find-what")`, so to add context to it we use `self.$("#find-what")`

**Review by swmitra — APPROVED** [review]
LGTM 👍

**Comment by swmitra** [comment]
@petetnt Once you approve, I will merge this PR 😄

**Review by petetnt — APPROVED** [review]
LGTM codewise, haven't had the time to retest the functionality though.

**Comment by swmitra** [comment]
@saurabh95 Great work :+1

**Comment by swmitra** [comment]
Merging...

**Pull request merged**